app/models/odk/response_parser.rb

```diff
@@ -0,0 +1,279 @@
+# frozen_string_literal: true
+
+module Odk
+  # Takes a response and odk data. Parses odk data into answer tree for response.
+  # Returns the response because in the case of multimedia, the parser may replace
+  # the response object the controller passes in with an
+  # existing response from the database.
+  class ResponseParser
+    attr_accessor :response, :raw_odk_xml, :files, :awaiting_media, :odk_hash
+
+    def initialize(response: nil, files: nil, awaiting_media: false)
+      raise "Submissions must have a mission" if response.mission.nil?
+      @response = response
+      @raw_odk_xml = files.delete(:xml_submission_file).read
+      @files = files
+      @response.source = "odk"
+      @awaiting_media = awaiting_media
+    end
+
+    # populates response tree and saves
+    def populate_response
+      process_xml(raw_odk_xml)
+      response
+    end
+
+    private
+
+    # Generates and saves a hash of the complete XML so that multi-chunk media form submissions
+    # can be uniquely identified and handled
+    def calculate_odk_hash
+      odk_hash || Digest::SHA256.base64digest(raw_odk_xml)
+    end
+
+    def process_xml(raw_odk_xml)
+      data = Nokogiri::XML(raw_odk_xml).root
+      lookup_and_check_form(id: data["id"], version: data["version"])
+      if existing_response
+        add_media_to_existing_response
+      else
+        response.odk_hash = awaiting_media ? calculate_odk_hash : nil
+        build_answer_tree(data)
+        response.associate_tree(response.root_node)
+      end
+    end
+
+    def add_media_to_existing_response
+      candidate_answers = response.answers.select { |a| a.pending_file_name.present? }
+      candidate_answers.each do |a|
+        populate_multimedia_answer(a, a.pending_file_name, a.questioning.qtype_name)
+      end
+    end
+
+    def build_answer_tree(data)
+      response.root_node = AnswerGroup.new(
+        questioning_id: response.form.root_id,
+        new_rank: 0
+      )
+      add_level(data, response.root_node)
+    end
+
+    def add_level(xml_node, response_node)
+      xml_node.elements.each do |child|
+        unless node_is_odk_header?(child)
+          if node_is_ir_question?(child)
+            response.incomplete = child.content == "yes"
+          else
+            add_response_node(child, response_node)
+          end
+        end
+      end
+    end
+
+    def add_response_node(node, response_node)
+      form_item = form_item(node.name)
```

Comment:
Can you please add a comment like "We need to consult FormItems here because the ODK XML doesn't give us explicit cues about node type."